x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/action/GetApiKeyRequest.java

```diff
@@ -0,0 +1,150 @@
+
+    public GetApiKeyRequest(StreamInput in) throws IOException {
+        super(in);
+        realmName = in.readOptionalString();
+        userName = in.readOptionalString();
+        apiKeyId = in.readOptionalString();
+        apiKeyName = in.readOptionalString();
+    }
+
+    private GetApiKeyRequest(@Nullable String realmName, @Nullable String userName, @Nullable String apiKeyId,
+            @Nullable String apiKeyName) {
+        this.realmName = realmName;
+        this.userName = userName;
+        this.apiKeyId = apiKeyId;
+        this.apiKeyName = apiKeyName;
+    }
+
+    public String getRealmName() {
+        return realmName;
+    }
+
+    public String getUserName() {
+        return userName;
+    }
+
+    public String getApiKeyId() {
+        return apiKeyId;
+    }
+
+    public String getApiKeyName() {
+        return apiKeyName;
+    }
+
+    /**
+     * Creates get API key request for given realm name
+     * @param realmName realm name
+     * @return {@link GetApiKeyRequest}
+     */
+    public static GetApiKeyRequest usingRealmName(String realmName) {
+        return new GetApiKeyRequest(realmName, null, null, null);
+    }
+
+    /**
+     * Creates get API key request for given user name
+     * @param userName user name
+     * @return {@link GetApiKeyRequest}
+     */
+    public static GetApiKeyRequest usingUserName(String userName) {
+        return new GetApiKeyRequest(null, userName, null, null);
+    }
+
+    /**
+     * Creates get API key request for given realm and user name
+     * @param realmName realm name
+     * @param userName user name
+     * @return {@link GetApiKeyRequest}
+     */
+    public static GetApiKeyRequest usingRealmAndUserName(String realmName, String userName) {
+        return new GetApiKeyRequest(realmName, userName, null, null);
+    }
+
+    /**
+     * Creates get API key request for given api key id
+     * @param apiKeyId api key id
+     * @return {@link GetApiKeyRequest}
+     */
+    public static GetApiKeyRequest usingApiKeyId(String apiKeyId) {
+        return new GetApiKeyRequest(null, null, apiKeyId, null);
+    }
+
+    /**
+     * Creates get api key request for given api key name
+     * @param apiKeyName api key name
+     * @return {@link GetApiKeyRequest}
+     */
+    public static GetApiKeyRequest usingApiKeyName(String apiKeyName) {
+        return new GetApiKeyRequest(null, null, null, apiKeyName);
+    }
+
+    @Override
+    public ActionRequestValidationException validate() {
+        ActionRequestValidationException validationException = null;
+        if (Strings.hasText(realmName) == false && Strings.hasText(userName) == false && Strings.hasText(apiKeyId) == false
+                && Strings.hasText(apiKeyName) == false) {
+            validationException = addValidationError("One of [api key id, api key name, username, realm name] must be specified",
+                    validationException);
+        }
+        if (Strings.hasText(apiKeyId)) {
```

Comment:
I would suggest something like 
```java
if (Strings.hasText(apiKeyId) || Strings.hasText(apiKeyName){
     if (Strings.hasText(realmName) || Strings.hasText(userName)) {
         validationException = addValidationError("username or realm name must not be specified when the api key id or api key name is specified");
     }
}
if (Strings.hasText(apiKeyId) && Strings.hasText(apiKeyName){
    validationException = addValidationError("only one of [api key id, api key name] can be specified");
}
```
instead of the following two `if` so that we don't end up with duplicated almost similar error messages. I think that for a request that comes with both an api key id and a name, 

> username or realm name must not be specified when the api key id or api key name is specified. only one of [api key id, api key name] can be specified" 

is better wrt to pointing out how the next request should come compared to 

> api key name or username or realm name must not be specified when the api key id is specified. api key id or username or realm name must not be specified when the api key name is specified

But I'm just nitpicking here, feel free to disregard the suggestion :)